### PR TITLE
QR (and other bar) code reading search bar functionality

### PR DESCRIPTION
**Pull request by nirvn**
This PR implements QR (and other bar-type) code scanning / decoding functionality to QField through the search bar. Its UI/UX is quite simple: when clicking on the search bar button, a new QR code icon is now present embedded in the search text field:

When clicking on it, a new code reader panel opens. Users can then simply focus their camera on a QR/bar code to decode. A visual "flash" feedback  occurs as soon as QField successful decoded to string. The decoded string appears below the camera frame for users to review:

When satisfied, users simply click on the ( :heavy_check_mark: ) button to trigger a search against that string. For e.g. here, the QR code was a user identifier linked to several features in a project dataset:

Other cool things can be achieved here (e.g., encoding a lat,lon string into a QR code), sky's the limit. Well, the search bar functionality is really :wink: 

There is a flashlight to help scanning codes in low light conditions.

Fixes https://github.com/opengisch/QField/issues/1085

_A follow up PR will be opened to add code scanning / decoding in the feature form text edit widget._

### CODE REVIEW

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for ccc2bd4d9ed9f88e8ecf593b3182dab35c42ca63: [**arm64-android**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-3104-95e21f5defd3b6067a82b6f36d3fa2aaf0517c30-arm64-android.apk)
<details>
<summary>Other architectures</summary>

- [arm-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-3104-95e21f5defd3b6067a82b6f36d3fa2aaf0517c30-arm-android.apk)
- [x64-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-3104-95e21f5defd3b6067a82b6f36d3fa2aaf0517c30-x64-android.apk)
- [x86-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-3104-95e21f5defd3b6067a82b6f36d3fa2aaf0517c30-x86-android.apk)
</details>

File: CMakeLists.txt [hunk 1]
```diff
@@ -188,6 +188,9 @@ if(MSVC)
   add_link_options("$<$<NOT:$<CONFIG:Debug>>:/OPT:ICF>")
 endif()
 
+
+find_package(ZXing REQUIRED)
```
Comment (m-kuhn) [inline CMakeLists.txt:192]: Can you move that into src/core/CMakeLists.txt?

Comment (nirvn) [reply CMakeLists.txt:192]: Done.

**Comment by nirvn** [comment]
Also, for the viewers at home, iOS failure is expected. It'll all magically fix itself when we move the iOS builds to vcpkg.

File: src/core/qgismobileapp.cpp [hunk 2]
```diff
@@ -149,6 +149,8 @@
 #include <qgsproject.h>
 #include <qgsprojectstorage.h>
 #include <qgsprojectstorageregistry.h>
+
+#include <ZXing/ReadBarcode.h>
```
Comment (m-kuhn) [inline src/core/qgismobileapp.cpp:153]: Is this include used for anything?

File: src/app/CMakeLists.txt [hunk 3]
```diff
@@ -1,5 +1,11 @@
 find_package(${QT_PKG}QuickCompiler QUIET)
 
+if(NOT Qt5QuickCompiler_FOUND
+   OR CMAKE_BUILD_TYPE MATCHES Debug
+   OR CMAKE_BUILD_TYPE MATCHES RelWithDebInfo)
+  set(RESOURCES ${CMAKE_SOURCE_DIR}/src/qml/qml.qrc)
+endif()
```
Comment (m-kuhn) [inline src/app/CMakeLists.txt:7]: This looks unrelated

File: src/core/qgismobileapp.cpp (hunk 2 above)
Comment (nirvn) [reply src/core/qgismobileapp.cpp:153]: Oups, it was used to test things here, will remove.

File: src/app/CMakeLists.txt (hunk 3 above)
Comment (nirvn) [reply src/app/CMakeLists.txt:7]: Right, I need to remove this commit. We still need to find a way for QField to compile properly on linux without vcpkg.

File: src/core/barcodedecoder.h [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+  signals:
+    void decodedStringChanged();
+
+  private:
+    QImage mImage;
+    QString mDecodedString;
+};
+
+class BarcodeDecoderThread : public QThread
```
Comment (m-kuhn) [inline src/core/barcodedecoder.h:46]: Can we move that to a private header?

Comment (nirvn) [reply src/core/barcodedecoder.h:46]: Or we move the class into the BarcodeVideoFilter cpp file? It's only used there ATM.

File: src/qml/CMakeLists.txt [hunk 5]
```diff
@@ -1,16 +1,17 @@
 find_package(${QT_PKG}QuickCompiler QUIET)
 find_package(${QT_PKG} COMPONENTS Quick)
 
+file(GLOB_RECURSE QML_SOURCES "${CMAKE_CURRENT_SOURCE_DIR}/*.qml")
```
Comment (m-kuhn) [inline src/qml/CMakeLists.txt:4]: Can we merge https://github.com/opengisch/QField/pull/3106 and rebase?

Comment (nirvn) [reply src/qml/CMakeLists.txt:4]: @m-kuhn , sounds good, approved that PR just now.

**Comment by m-kuhn** [comment]
> Edit: just noticed ubuntu actually has the library (and the -dev counterpart) in its repository, it's probably not worth dragging the whole 200,000 lines of code in our source tree then.

Good, I think vendoring dependencies is a last resort only, we should refrain from doing so and stick to upstream

**Comment by nirvn** [comment]
And just like that, thanks to @m-kuhn Herculean efforts, iOS turns green :)

**Comment by m-kuhn** [comment]
Did you just add a 🔦 ???

**Comment by nirvn** [comment]
@m-kuhn , :)

This is ready for review and merging. I'll implement the feature form functionality in a different PR.

File: src/core/barcodedecoder.cpp [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+  if ( imageFormat != ZXing::ImageFormat::None )
+  {
+    ZXing::ImageView imageView( image.bits(), image.width(), image.height(), imageFormat, static_cast<int>( image.bytesPerLine() ) );
+
+    ZXing::DecodeHints hints;
+    hints.setFormats( ZXing::BarcodeFormat::Any );
+    hints.setTryRotate( false );
+
+    ZXing::Result result = ZXing::ReadBarcode( imageView, hints );
+    QString resultText = QString::fromWCharArray( result.text().c_str() );
```
Comment (m-kuhn) [inline src/core/barcodedecoder.cpp:72]:
```suggestion
    ZXing::Result result = ZXing::ReadBarcode( imageView, hints );
    const std::wstring text = result.text();
    QString resultText = QString::fromWCharArray( text.c_str() );
```

Just to be sure we really get the 16 bit return value

File: src/core/barcodedecoder.h [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+    Q_PROPERTY( QString decodedString READ decodedString WRITE setDecodedString NOTIFY decodedStringChanged )
+
+  public:
+    explicit BarcodeDecoder( QObject *parent = nullptr );
+
+    QString decodedString() const { return mDecodedString; }
+    void setDecodedString( const QString &decodedString );
+
+    void decodeImage( const QImage &image );
```
Comment (m-kuhn) [inline src/core/barcodedecoder.h:36]: Worth a test?

**Review by m-kuhn — COMMENTED** [review]
C++ part looks good, some minor things/questions

@suricactus could you have an eye on the QML parts?

**Comment by nirvn** [comment]
@m-kuhn , thanks for the review, comments addressed. As for the :flashlight: , it's implemented, but my phone doesn't seem to register its flash as a torch. Either that or we're facing another 5.14 bug.

**Comment by mbernasocchi** [comment]
@nirvn @m-kuhn is ther a way to trigger another apks build? :)

File: src/core/barcodedecoder.cpp [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    hints.setTryRotate( true );
+
+    ZXing::Result result = ZXing::ReadBarcode( imageView, hints );
+    const std::wstring text = result.text();
+    QString resultText = QString::fromWCharArray( text.c_str() );
+    if ( !resultText.isEmpty() && mDecodedString != resultText )
+    {
+      mDecodedString = resultText;
+      emit decodedStringChanged();
+    }
```
Comment (suricactus) [inline src/core/barcodedecoder.cpp:78]: If I get it right, it means if I have `"scanned barcode"` already set from the previous scan and I scan again and I scan empty string `""`, there will be `decodedStringChanged()`, therefore the UI would not update?

Is it on purpose?

Comment (nirvn) [reply src/core/barcodedecoder.cpp:78]: Correct.

File: src/core/barcodevideofilter.cpp [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+    {
+    }
+
+  private:
+    void run() override
+    {
+      if ( mDecoder )
+      {
+        mDecoder->decodeImage( mImage );
+      }
```
Comment (suricactus) [inline src/core/barcodevideofilter.cpp:38]: Just wondering how many calls it has per second? If they are too much, we can drain the battery like crazy, probably a small sleep makes sense, so we do 1-2 scans per second max?

File: src/core/barcodevideofilter.h [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+  public:
+    explicit BarcodeVideoFilter();
+
+    BarcodeDecoder *decoder() const { return mDecoder; }
+    void setDecoder( BarcodeDecoder *decoder );
+
+    void decodeVideoFrame( QVideoFrame *input );
+    bool isDecoding() const { return mDecodingThread; }
+
+    QVideoFilterRunnable *createFilterRunnable() override;
```
Comment (suricactus) [inline src/core/barcodevideofilter.h:41]: A bit of docs would be great.

File: src/qml/qgismobileapp.qml [hunk 11]
```diff
@@ -2994,6 +2994,22 @@ ApplicationWindow {
       }
   }
 
+  MouseArea {
+    id: barcodeReaderCatcher
+    anchors.fill: parent
+    enabled: barcodeReader.visible
+
+    onClicked: {
+      // Needed to avoid people modifying context within which the code reader was called
```
Comment (suricactus) [inline src/qml/qgismobileapp.qml:3003]: Sorry, it is not clear what context you are talking about and how the context could be modified?

File: src/qml/BarcodeReader.qml [hunk 12]
```diff
@@ -0,0 +1,267 @@
...
+import org.qfield 1.0
+
+import Theme 1.0
+
+Popup {
+  id : barcodeReader
+
+  signal decoded(var string)
+
+  property int popupWidth: mainWindow.width <= mainWindow.height ? mainWindow.width - 40 : mainWindow.height - 40
```
Comment (suricactus) [inline src/qml/BarcodeReader.qml:16]: While not sure if it is worth it for this PR, I believe we should make all this magic numbers Theme variables.   In this case is `popupScreenEdgeMargin` I guess.

File: src/qml/BarcodeReader.qml [hunk 13]
```diff
@@ -0,0 +1,267 @@
...
+            strokeColor: "white"
+            strokeStyle: ShapePath.SolidLine
+            joinStyle: ShapePath.MiterJoin
+            fillColor: "transparent"
+
+            startX: 20
+            startY: 60
+            PathLine { x: 20; y: 25 }
+            PathArc { x: 25; y: 20; radiusX: 5; radiusY: 5 }
+            PathLine { x: 60; y: 20 }
```
Comment (suricactus) [inline src/qml/BarcodeReader.qml:179]: I hope we never have to change these values, as we have to recalculate them manually. Other option would be to make them dynamically calculated like:
```
ShapePath {
    startX: 20
    startY: 60
    property int borderRadius:  5
    PathLine { x: parent.startX; y: parent.startX + parent.borderRadius }
    // ...
 }
```
No strong opinion here.

**Review by suricactus — APPROVED** [review]
Fantastic work! :iphone: 📓

File: src/core/barcodevideofilter.cpp (hunk 9 above)
Comment (nirvn) [reply src/core/barcodevideofilter.cpp:38]: I've discussed this with @suricactus yesterday. To leave a record of our discussion: capping decoding to say 2 scans per second would probably save some battery (not that much as the camera is what sucks battery out), but it risks having situations where the environment isn't great (shacky hands like mine, shacky barcode, or both) and we'd hit "a bad frame" twice per second and end up having to wait a long longer for the barcode to be decoded.  When decoding as much as possible, we maximize chance of success.

Hence thinking we should go for the decode-as-many-frames-as-possible to begin with.

File: src/qml/BarcodeReader.qml (hunk 12 above)
Comment (nirvn) [reply src/qml/BarcodeReader.qml:16]: @suricactus , very good idea sir.

**Comment by nirvn** [comment]
@suricactus , comments addressed, all very useful.
